<launch>

  <arg name="ticks_per_meter" value="8484" />
  <arg name="wheel_separation" value="0.22" />
  <arg name="max_motor_speed" value="4000" />

  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find rover)/description/rover_reversed.urdf.xml"/>

  <!-- Publish the robot state -->
  <node name="robot_state_publisher" pkg="robot_state_publisher"
        type="robot_state_publisher">
    <param name="publish_frequency" value="10.0"/>
  </node>

  <!-- Provide simulated control of the robot joint angles -->
  <node name="joint_state_publisher" pkg="joint_state_publisher"
        type="joint_state_publisher">
    <param name="use_gui" value="False" />
    <param name="rate" value="10.0"/>
  </node>
  
  <node pkg="tf" type="static_transform_publisher" name="imu_link_broadcaster" args="0 0 0 0 0 0 imu_base imu_link 100" />

  <node name="controller" pkg="diff_drive" type="diff_drive_controller"
        output="screen">
    <rosparam subst_value="true">
      ticks_per_meter: $(arg ticks_per_meter)
      wheel_separation: $(arg wheel_separation)
      max_motor_speed: $(arg max_motor_speed)
      timeout: 1.0
    </rosparam>
    <remap from="cmd_vel" to="/robot/cmd_vel" />
  </node>
  
  <node name="odom_publisher" pkg="diff_drive" type="diff_drive_odometry"
        output="screen">
    <rosparam subst_value="true">
      ticks_per_meter: $(arg ticks_per_meter)
      wheel_separation: $(arg wheel_separation)
      encoder_min: -32768
      encoder_max: 32768
      publish_tf: false
    </rosparam>
  </node>

  <node name="rover_controller" pkg="rover" type="rover_controller.py" output="screen">
    <rosparam subst_value="true">
      max_motor_speed: $(arg max_motor_speed)
      reverse_motor_direction: true
    </rosparam>
    <remap from="~lwheel_desired_rate" to="lwheel_desired_rate" />
    <remap from="~rwheel_desired_rate" to="rwheel_desired_rate" />
    <remap from="~lwheel_ticks" to="lwheel_ticks" />
    <remap from="~rwheel_ticks" to="rwheel_ticks" />
  </node>

  <node name="diff_drive_go_to_goal" pkg="diff_drive" type="diff_drive_go_to_goal" output="screen">
    <param name="~rate" value="20" />
    <param name="~kP" value="0.5" />
    <param name="~kA" value="1.0" />
    <param name="~kB" value="-0.8" />
    <param name="~max_linear_speed" value="0.2" />
    <param name="~min_linear_speed" value="0.05" />
    <param name="~max_angular_speed" value="0.7" />
    <param name="~min_angular_speed" value="0.1" />
    <param name="~linear_tolerance" value="0.1" />
    <param name="~angular_tolerance" value="0.087" />
    <param name="~forwardMovementOnly" value="true" />
    <remap from="cmd_vel" to="/robot/cmd_vel" />
  </node>
  
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization_odom" clear_params="true"> 
    <rosparam command="load" file="$(find rover)/config/robot_localization.yaml" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find rover)/config/driver.rviz" />

</launch>
